Platform agnostic file functions

mcostalba opened the pull request:
Reduce the platform specific code to bare minimum
and localize all in a single place.

lucasart commented:
Hum... I don't understand what the compiler is complaining about:
```
g++ -Wall -Wcast-qual -fno-exceptions -fno-rtti -std=c++11  -pedantic -Wextra -Wshadow -m64 -DNDEBUG -O3 -DIS_64BIT -msse -msse3 -mpopcnt -DUSE_POPCNT -DUSE_PEXT -mbmi -mbmi2 -flto   -c -o syzygy/tbprobe.o syzygy/tbprobe.cpp
In file included from syzygy/tbprobe.cpp:20:0:
syzygy/tbcore.cpp: In member function ‘char* TBFile::map(uint64_t*)’:
syzygy/tbcore.cpp:77:42: error: invalid conversion from ‘int’ to ‘std::ios_base::openmode {aka std::_Ios_Openmode}’ [-fpermissive]
     int fd = open(fname.c_str(), O_RDONLY);
                                          ^
```

mcostalba commented:
It seems a name resolution issue, fstream has a member called 'open', try
this:

int fd = ::open(fname.c_str(), O_RDONLY);


Oddly enough I don't get the compile error on my platform.



On Tue, Apr 19, 2016 at 4:33 PM, lucasart <notifications@github.com> wrote:

> Hum... I don't understand what the compiler is complaining about:
>
> g++ -Wall -Wcast-qual -fno-exceptions -fno-rtti -std=c++11  -pedantic -Wextra -Wshadow -m64 -DNDEBUG -O3 -DIS_64BIT -msse -msse3 -mpopcnt -DUSE_POPCNT -DUSE_PEXT -mbmi -mbmi2 -flto   -c -o syzygy/tbprobe.o syzygy/tbprobe.cpp
> In file included from syzygy/tbprobe.cpp:20:0:
> syzygy/tbcore.cpp: In member function ‘char* TBFile::map(uint64_t*)’:
> syzygy/tbcore.cpp:77:42: error: invalid conversion from ‘int’ to ‘std::ios_base::openmode {aka std::_Ios_Openmode}’ [-fpermissive]
>      int fd = open(fname.c_str(), O_RDONLY);
>                                           ^
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/official-stockfish/Stockfish/pull/643#issuecomment-211948169>
>

stuwph commented:
syzygy access seems to be broken, even after 'additional formatting' 
(windows x64 mingw 5.30 compile)

stuwph commented:
still no syzygy after 'assorted code reshuffling' under windows fritz15 and arena351 gui
initializing the tables, but while accessing in play throwing 'corrupted table'  (arena)

stuwph commented:
worked until  'Platform agnostic file functions'

lucasart commented on src/syzygy/tbcore.cpp:
```diff
@@ -466,21 +226,125 @@ static const short KK_idx[10][64] = {
+
+        while (std::getline(ss, path, SEP_CHAR)) {
+            std::ifstream::open(path + "/" + fname);
+            if (is_open())
```
before `return`, you need to set `fname = path + "/" + fname`

lucasart commented on src/syzygy/tbcore.cpp:
```diff
@@ -466,21 +226,125 @@ static const short KK_idx[10][64] = {
+        fstat(fd, &statbuf);
+        *mapping = statbuf.st_size;
+        char* data = (char*)mmap(NULL, statbuf.st_size, PROT_READ, MAP_SHARED, fd, 0);
+        close(fd);
```
Same as `open()`, you need `::close(fd)`

lucasart commented:
@mcostalba: i managed to compile it, and fix the obvious. see comments in line. but it is still broken in a more subtle way now... wrong syzygy signature.

mcostalba commented:
@lucasart I have fixed the issues and updated the pull request. Can you please confirm that is broken on [this](https://github.com/official-stockfish/Stockfish/pull/643/commits/0090c72e356314da) commit?

mcostalba commented:
I have finally found why I crash on sygyzy.epd, here there are some illegal mate positions:

K7/8/1q6/8/6k1/8/8/8 w - - 71
6b1/8/8/2b5/8/7K/6Q1/6k1 b - - 46
k2q1Kr1/4n3/8/3b4/8/8/8/8 w - - 83
8/7k/8/8/R7/8/2K4Q/1R4Q1 b - - 5
3K4/1b3q2/3b4/8/2k5/2P5/8/8 w - - 43

And there are many more, without these I am able to run the bench with TB.

@lucasart how you are able to run the bench on those? each of these positions fail on SF even without TB !

mcostalba commented:
OK, now that I am able to run 'bench' I have finally found the bug: PR fixed and ready to be merged.

lucasart commented:
@mcostalba: Ah, I fixed my random position generator to filter mates and stalemates, but did not update `syzygy.epd`... I will open a pull request to update `syzygy.epd`. SF can run on these positions, but not in debug mode (assert crash).

lucasart commented:
@mcostalba: patch is ok to merge.

stuwph commented:
getting a mingw 5.3 linker error since 'Improve TBFile API'
 
syzygy/tbprobe.o:tbprobe.cpp:(.text.unlikely+0x24e7): undefined reference to `(anonymous namespace)::WDLEntry::TB_MAGIC'
syzygy/tbprobe.o:tbprobe.cpp:(.text.unlikely+0x3472): undefined reference to `(anonymous namespace)::DTZEntry::TB_MAGIC'

mcostalba commented:
Which version of gcc you have?

Type g++ - v


On Thursday, April 28, 2016, stuwph <notifications@github.com> wrote:

> getting a mingw 5.3 linker error since 'Improve TBFile API'
>
> syzygy/tbprobe.o:tbprobe.cpp:(.text.unlikely+0x24e7): undefined reference
> to (anonymous namespace)::WDLEntry::TB_MAGIC'
> syzygy/tbprobe.o:tbprobe.cpp:(.text.unlikely+0x3472): undefined reference
> to(anonymous namespace)::DTZEntry::TB_MAGIC'
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly or view it on GitHub
> <https://github.com/official-stockfish/Stockfish/pull/643#issuecomment-215313306>
>

stuwph commented:
thread model: posix
gcc version 5.3.0 (Rev5, Built by MSYS2 project)

mcostalba commented:
Ok, I have pushed a possible fix, could you please verify now it works?


On Thu, Apr 28, 2016 at 8:04 AM, stuwph <notifications@github.com> wrote:

> thread model: posix
> gcc version 5.3.0 (Rev5, Built by MSYS2 project)
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly or view it on GitHub
> <https://github.com/official-stockfish/Stockfish/pull/643#issuecomment-215318025>
>

stuwph commented:
works again :)
